scheds/include/scx/sdt_task_impl.bpf.h

```diff
@@ -0,0 +1,579 @@
+
+		chunk = desc->chunk;
+		cast_kern(chunk);
+
+		desc_children = (struct sdt_task_desc * __arena *)chunk->descs;
+		desc = desc_children[pos];
+
+		if (!desc) {
+			bpf_spin_unlock(&sdt_task_lock);
+			bpf_printk("freeing nonexistent idx [0x%lx] (level %d)", idx, level);
+			return;
+		}
+	}
+
+	cast_kern(desc);
+
+	chunk = desc->chunk;
+	cast_kern(chunk);
+
+	pos = idx & mask;
+	data = chunk->data[pos];
+	if (!data) {
+		bpf_spin_unlock(&sdt_task_lock);
+		bpf_printk("freeing idx [0x%lx] (%p) without data", idx, &chunk->data[pos]);
+		return;
+	}
+
+	cast_kern(data);
+
+	data[pos] = (struct sdt_task_data) {
+		.tid.gen = data->tid.gen + 1,
+		.tptr = 0,
+	};
+
+	/* Zero out one word at a time. */
+	bpf_for(i, 0, sdt_task_data_pool.free_size / 8) {
+		data->payload[i] = 0;
+	}
+
+	bpf_for(u, 0, SDT_TASK_LEVELS) {
+		level = SDT_TASK_LEVELS - 1 - u;
+
+		/* Only propagate upwards if we are the parent's only free chunk. */
+		desc = lv_desc[level];
+
+		sdt_set_idx_state(desc, lv_pos[level], false);
+
+		cast_kern(desc);
+
+		desc->nr_free += 1;
+		if (desc->nr_free > 1)
+			break;
+	}
+
+	bpf_spin_unlock(&sdt_task_lock);
+
+	return;
+}
+
+static SDT_TASK_FN_ATTRS
+void __arena *sdt_task_retrieve(struct task_struct *p)
+{
+	struct sdt_task_data __arena *data;
+	struct sdt_task_map_val *mval;
+
+	sdt_arena_verify();
+
+	mval = bpf_task_storage_get(&sdt_task_map, p, 0, 0);
+	if (!mval)
+		return NULL;
+
+	data = mval->data;
+
+	return (void __arena *)data->payload;
+}
+
+
+static SDT_TASK_FN_ATTRS void sdt_task_free(struct task_struct *p)
+{
+	struct sdt_task_map_val *mval;
+
+	sdt_arena_verify();
+
+	mval = bpf_task_storage_get(&sdt_task_map, p, 0, 0);
+	if (!mval)
+		return;
+
+	sdt_task_free_idx(mval->tid.idx);
+	mval->data = NULL;
+}
+
+
+static SDT_TASK_FN_ATTRS
+int sdt_task_find_empty(struct sdt_task_desc __arena *desc, struct sdt_task_desc * __arena *descp, __u64 *idxp)
+{
+	struct sdt_task_desc * __arena *desc_children,  __arena *new_chunk;
+	struct sdt_task_desc __arena *lv_desc[SDT_TASK_LEVELS];
+	struct sdt_task_chunk __arena *chunk;
+	struct sdt_task_desc __arena *tmp;
+	__u64 lv_pos[SDT_TASK_LEVELS];
+	__u64 u, pos, level;
+	__u64 idx = 0;
+
+	bpf_for(level, 0, SDT_TASK_LEVELS) {
+		pos = sdt_chunk_find_empty(desc);
+
+		/* Something has gon terribly wrong. */
+		if (pos > SDT_TASK_ENTS_PER_CHUNK)
+			return -EINVAL;
+
+		if (pos == SDT_TASK_ENTS_PER_CHUNK)
+			return -ENOMEM;
+
+		idx <<= SDT_TASK_ENTS_PER_CHUNK_SHIFT;
+		idx |= pos;
+
+		/* Log the levels to complete allocation. */
+		lv_desc[level] = desc;
+		lv_pos[level] = pos;
+
+		/* The rest of the loop is for internal node traversal. */
+		if (level == SDT_TASK_LEVELS - 1)
+			break;
+
+		cast_kern(desc);
+
+		chunk = desc->chunk;
+		cast_kern(chunk);
+
+		desc_children = (struct sdt_task_desc * __arena *)chunk->descs;
+		desc = desc_children[pos];
+
+		/* Someone else is populating the subtree. */
+		if (desc == (void *)SDT_TASK_ALLOC_RESERVE)
+			return -EAGAIN;
```

Comment:
Wouldn't this make everyone else who's racing busy loop on the same slot over and over again? The allocation can take a substantial amount of time if it involves reclaim.